File: src/previous/day2/Queue.ts
// A >> B >> C >> D
// First in last out
type Node<T> = {
    value: T
    next?: Node<T>
}
function createNode<T>(value: T): Node<T> {
    return { value }
}

export default class Queue<T> {
    public length: number
    private head?: Node<T>
    private tail?: Node<T>

    constructor() {
        this.length = 0
        this.head = this.tail = undefined
    }

    enqueue(item: T): void {
        this.length++
        const node = createNode(item)
        if (!this.tail) {
            this.head = this.tail = node
            return
        }
        this.tail.next = node
        this.tail = node
    }
    deque(): T | undefined {
        if (!this.head) {
            return undefined
        }

        this.length--
        // if there is only one item in the list, after the head is popped out,
        // tail will still be pointing to the popped out node, set it to undefined
        // and leave the head to walk to the next and be undefined
        if (!this.head.next) {
            this.tail = undefined
        }
        const out = this.head.value
        this.head = this.head.next

        return out
    }
    peek(): T | undefined {
        return this.head?.value
    }
}

File: go/day01/Queue/Queue.go
package queue

type Node[T comparable] struct {
	next  *Node[T]
	value T
}

type Queue[T comparable] struct {
	head   *Node[T]
	tail   *Node[T]
	length int
}

func MakeQueue[T comparable]() *Queue[T] {
	return &Queue[T]{
		length: 0,
	}
}

func (q *Queue[T]) Enque(item T) error {
	node := &Node[T]{
		value: item,
	}

	q.length++
	if q.tail == nil {
		q.head = node
		q.tail = node
		return nil
	}

	q.tail.next = node
	q.tail = node

	return nil
}

func (q *Queue[T]) Deque() (T, error) {
	var out T
	if q.head == nil {
		return out, nil
	}

	out = q.head.value
	q.length--
	if q.length == 0 {
		q.head = nil
		q.tail = nil
		return out, nil
	}

	q.head = q.head.next
	return out, nil
}

func (q *Queue[T]) Peek() T {
	out := q.head.value
	return out
}
